### PR TITLE
Replace Javascript front-end code with Typescript

### LINKED ISSUE #40: Replace Javascript front-end code with Typescript
The front-end of the electron app is written in vanilla javascript and jquery.

It's rather ugly but it works. Many people advised that rewritting in Typescript will help improve it and make it more maintainable.

### PR DESCRIPTION
**Pull request by StevenJNPearce**
Fixes #40 

I haven't tested this yet as I've been developing on Windows. Tomorrow I will get a Linux desktop running to check everything is working as intended.

### CODE REVIEW

**Comment by LefterisJP** [comment]
Hey @StevenJNPearce please rebase on latest master as you seem to have based this off from an older version.

Also yeah since Windows is not supported (yet) you will need to test it from a linux machine.

**Comment by StevenJNPearce** [comment]
@LefterisJP The dropdown on the "Trades" button of the Left navigation menu is not working. I'm having trouble diagnosing the issue because I can't find the code for this dropdown anywhere - it is not in navigation.js
Maybe you can take a look to try and resolve this issue?

**Comment by StevenJNPearce** [comment]
Sorry about the lack of update, had some other work to attend to. By not working I mean clicking the trades button wasn't doing anything. I will rebase this against the new commits and find out what the problem is + fix.

**Comment by LefterisJP** [comment]
Thank you. I will review, but the UI tests are still failing as you can see here: https://travis-ci.org/rotkehlchenio/rotkehlchen/jobs/414359871

**Comment by vs77bb** [comment]
Hi @StevenJNPearce were you able to review @LefterisJP's comments here? Hope you're doing well :+1:

**Comment by StevenJNPearce** [comment]
I'd put this to one side due to @LefterisJP 's vacation, will get back on it :)

Odd the issue didn't occur with the CI build, not sure of the cause. Actually it doesn't matter; I can see how to resolve it.
`types/jqueryui@1.12.2` is a dependency of `types/jqueryui.datetimepicker` but It looks like I've included this accidentally - the dependecy is `jquery.datetimepicker` and not `jqueryui.datetimepicker`

**Comment by LefterisJP** [comment]
@StevenJNPearce I am back from vacation. Can you fix the dependencies then so I can try testing again?

**Comment by StevenJNPearce** [comment]
@LefterisJP Will do.

**Comment by LefterisJP** [comment]
I rebased your branch on latest master and also removed package-lock.json. Please don't commit it. I will review the code now.

**Comment by LefterisJP** [comment]
You have commited a file called `git`. Please remove it by rebasing it away. I don't want it in the history.

File: ui/asset_table.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+            $(str).appendTo($('#' + placement_id));
+        } else if (placement_type === 'insertAfter') {
+            $(str).insertAfter('#' + placement_id);
+        } else {
+            let err = new Error();
+            throw_with_trace('Invalid AssetTable construction value for placement_type: ' + placement_type);
+        }
+        this.populate(table_data, draw_cb);
+    }
+    format_data(original_data: any): Array<any> {
```
Comment (LefterisJP) [inline ui/asset_table.ts]: Should the `original_data` also be `Array<any>` here?

File: ui/asset_table.ts [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+                parseFloat(original_data[asset]['usd_value'])
+                .toFixed(settings.floating_precision);
+                let row = {[this.first_column_name]: asset, 'amount': amount, 'usd_value': value};
+                data.push(row);
+            }
+        }
+        return data;
+    }
+    
+    populate(table_data: any, draw_cb: any): void {
```
Comment (LefterisJP) [inline ui/asset_table.ts]: Same comment as above for `table_data` being an array and `draw_cb` being a callable

File: ui/asset_table.ts [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        let first_column_name = this.first_column_name;
+        let init_obj = {
+            'data': data,
+            'columns': [
+                {'data': first_column_name, 'title': string_capitalize(first_column_name)},
+                {'data': 'amount', 'title': 'Amount'},
+                {
+                    'data': 'usd_value',
+                    'title': settings.main_currency.ticker_symbol + ' value',
+                    'render': function (data: any , type: any, row: any) {
```
Comment (LefterisJP) [inline ui/asset_table.ts]: Please add proper types here for `row`.

File: ui/balances_table.ts [hunk 4]
```diff
@@ -89,11 +96,11 @@ function total_table_recreate() {
     update_last_balance_save();
 }
 
-function total_table_add_balances(location, query_result) {
+export function total_table_add_balances(location: string, query_result: any) {
```
Comment (LefterisJP) [inline ui/balances_table.ts]: Here `query_result` should be an object. It's value is something like `{'USD': '100', 'BTC': '1'}`

File: ui/dashboard.ts [hunk 5]
```diff
@@ -1,25 +1,24 @@
...
+        let target_location = determine_location(this.getAttribute('href'));
         if (target_location.startsWith('exchange_')) {
             let exchange_name = target_location.substring(9);
             assert_exchange_exists(exchange_name);
-            console.log("Going to exchange " + exchange_name);
+            console.log('Going to exchange ' + exchange_name);
             create_or_reload_exchange(exchange_name);
         } else {
-            throw "Invalid link location " + target_location;
+            throw new Error('Invalid link location ' + target_location);
```
Comment (LefterisJP) [inline ui/dashboard.ts]: What's the difference here? What does `new Error('XXX')` do compared to a simple throw?

File: ui/dashboard.ts [hunk 6]
```diff
@@ -156,16 +155,16 @@ ipc.on('failed', (event, message) => {
...
     // send ack to main.js
     ipc.send('ack', 1);
 });
 
 
-function init_dashboard() {
+export function init_dashboard() {
     // add callbacks for dashboard to the monitor
-    monitor_add_callback('query_exchange_balances', function (result) {
+    monitor_add_callback('query_exchange_balances', function (result: any) {
```
Comment (LefterisJP) [inline ui/dashboard.ts]: Here result is also an object

File: ui/dashboard.ts [hunk 7]
```diff
@@ -182,8 +181,8 @@ function init_dashboard() {
         );
         total_table_add_balances(result['name'], result['balances']);
     });
-    monitor_add_callback('query_blockchain_balances', function (result) {
-        if (result['message'] != '') {
+    monitor_add_callback('query_blockchain_balances', function (result: any) {
```
Comment (LefterisJP) [inline ui/dashboard.ts]: same here

File: ui/definitions.d.ts [hunk 8]
```diff
@@ -0,0 +1,14 @@
+interface JQueryStatic {
```
Comment (LefterisJP) [inline ui/definitions.d.ts]: What does this file do?

File: ui/index.html [hunk 9]
```diff
@@ -163,6 +163,9 @@ <h1 class="page-header">Dashboard</h1>
 	<!-- Include jquery the electron way -->
 	<script>
     window.jQuery = window.$ = require('jquery');
+    require( 'datatables.net' )( window, $ );
```
Comment (LefterisJP) [inline ui/index.html]: Can you please remove the extra space between the parentheses and its content here and below?

File: ui/index.html [hunk 10]
```diff
@@ -163,6 +163,9 @@ <h1 class="page-header">Dashboard</h1>
...
+    require( 'datatables.net-dt' )( window, $ );
```
Comment (LefterisJP) [inline ui/index.html]: Also why did you add these two requires here? What do they achieve?

File: ui/monitor.ts [hunk 11]
```diff
@@ -73,15 +79,15 @@ function remove_task(task_id) {
         }
     }
     delete tasks_map[task_id];
-    $('.task'+task_id).remove();
+    $('.task' + task_id).remove();
 }
 
-function monitor_add_callback(task_type, cb) {
+export function monitor_add_callback(task_type: string, cb: any) {
```
Comment (LefterisJP) [inline ui/monitor.ts]: `cb` here is a callback. Can you specify it with a callable type?

File: ui/monitor.ts [hunk 12]
```diff
@@ -90,28 +96,31 @@ function monitor_tasks() {
...
-    for (var task_id in tasks_map) {
+    for (let task_id in tasks_map) {
         let task = tasks_map[task_id];
         if (task.id == null) {
             console.log('NULL TASK ID: ' + JSON.stringify(task, null, 4));
             continue;
         }
 
-        client.invoke("query_task_result", task.id, function (error, res) {
+        client.invoke('query_task_result', task.id, function (error: Error, res: any) {
```
Comment (LefterisJP) [inline ui/monitor.ts]: here `res` should be an object and `error` a string, right?

File: ui/navigation.ts [hunk 13]
```diff
@@ -0,0 +1,109 @@
+import { add_settings_listeners, assert_exchange_exists, create_settings_ui, settings } from './settings';
+import { add_accounting_settings_listeners, create_accounting_settings } from './accounting_settings';
+import { add_taxreport_listeners, create_taxreport_ui } from './taxreport';
+import { add_otctrades_listeners, create_otctrades_ui } from './otctrades';
+import { create_or_reload_dashboard } from './dashboard';
+import { add_user_settings_listeners, create_user_settings } from './user_settings';
+
+export function determine_location(url: any) {
```
Comment (LefterisJP) [inline ui/navigation.ts]: isn't `url` a string?

File: ui/navigation.ts [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+        settings.page_accounting_settings = $('#page-wrapper').html();
+    } else if (settings.current_location === 'taxreport') {
+        console.log('Saving tax report ...');
+        settings.page_taxreport = $('#page-wrapper').html();
+    } else {
+        throw new Error('Invalid link location ' + settings.current_location);
+    }
+}
+
+export function change_location(target: any) {
```
Comment (LefterisJP) [inline ui/navigation.ts]: `target` should be a string

File: ui/navigation.ts [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+    save_current_location();
+    console.log('Changing location to ' + target);
+    settings.current_location = target;
+}
+
+export function create_or_reload_page(name: string, create_callback: any, always_callback: any) {
```
Comment (LefterisJP) [inline ui/navigation.ts]: please add the types for the callbacks

File: ui/navigation.ts [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+    } else {
+        console.log(`At create/reload ${name} with a populated page index`);
+        $('#page-wrapper').html(settings['page_' + name]);
+    }
+    always_callback();
+}
+
+export function init_navigation() {
+    $('#side-menu a').click(function(event: any) {
+       // event.preventDefault();
```
Comment (LefterisJP) [inline ui/navigation.ts]: Why are you commenting out the `event.preventDefault()` line here?

File: ui/otctrades.ts [hunk 17]
```diff
@@ -1,34 +1,50 @@
...
+
+let CURRENT_TRADE = {} as TradeData;
+
+function update_valhtml(selector_text: any, val: any) {
     $(selector_text).val(val);
     $(selector_text).html(val);
 }
 
-function format (d) {
+function format (d: any) {
```
Comment (LefterisJP) [inline ui/otctrades.ts]: d should be an object

File: ui/otctrades.ts [hunk 18]
```diff
@@ -1,34 +1,50 @@
...
+import client from './zerorpc_client';
 
-var OTC_TRADES_TABLE = null;
-var CURRENT_TRADE = {};
 
-function update_valhtml(selector_text, val) {
+let OTC_TRADES_TABLE = null as DataTables.Api;
+
+interface TradeData {
+    id: any;
```
Comment (LefterisJP) [inline ui/otctrades.ts]: If I am not mistake id is a number. Either that or a string.

File: ui/otctrades.ts [hunk 19]
```diff
@@ -1,34 +1,50 @@
...
+import { form_button, form_entry, form_select, form_text  } from './elements';
+import client from './zerorpc_client';
 
-var OTC_TRADES_TABLE = null;
-var CURRENT_TRADE = {};
 
-function update_valhtml(selector_text, val) {
+let OTC_TRADES_TABLE = null as DataTables.Api;
+
+interface TradeData {
```
Comment (LefterisJP) [inline ui/otctrades.ts]: Since I am just starting to understand typescript, what does the interface here do? Why not a class? It's just defining what is inside `CURRENT_TRADE` ?

File: ui/renderer.ts [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+import { init_navigation } from './navigation';
+import { init_taxreport } from './taxreport';
+client.connect('tcp://127.0.0.1:4242');
+
+function create_page_header(title: string) {
+    return '<div class="row"><div class="col-lg-12"><h1 class=page-header">' + title + '</h1></div></div>';
+}
+
+
+function create_currency_dropdown(fa_default_icon: any) {
```
Comment (LefterisJP) [inline ui/renderer.ts]: `fa_default_icon` should be a string here

File: ui/settings.ts [hunk 21]
```diff
@@ -0,0 +1,211 @@
...
+    );
+    $(str).appendTo($('.panel-body'));
+
+    str = form_button('Save', 'settingssubmit');
+    str += '</form></div>';
+    $(str).appendTo($('.panel-body'));
+}
+
+interface Settings {
+    usd_to_fiat_exchange_rates: any;
```
Comment (LefterisJP) [inline ui/settings.ts]: This is an object with exchange rates. i.e.: `{'EUR': 1.232, 'CNY': 200.2}` e.t.c.

File: ui/settings.ts [hunk 22]
```diff
@@ -0,0 +1,211 @@
...
+    page_exchange: any;
+    datetime_format: string;
+    has_premium: boolean;
+    premium_should_sync: boolean;
+    start_suggestion: string;
+    eth_rpc_port: string;
+    balance_save_frequency: number;
+    include_crypto2crypto: boolean;
+    taxfree_after_period: number;
+    last_balance_save: any;
```
Comment (LefterisJP) [inline ui/settings.ts]: this is a unix timestamp, so a number

File: ui/topmenu.ts [hunk 23]
```diff
@@ -0,0 +1,46 @@
+
+
```
Comment (LefterisJP) [inline ui/topmenu.ts]: please remove the extra lines in the top of the file here

File: ui/user_settings.ts [hunk 24]
```diff
@@ -24,56 +27,56 @@ const ExchangeBadge = ({ name, css_class }) => `
 </div>
 `;
 
-function show_loading(selector) {
+function show_loading(selector?: any) {
```
Comment (LefterisJP) [inline ui/user_settings.ts]: isn't a jquery selector a `string` here?

File: ui/user_settings.ts [hunk 25]
```diff
@@ -24,56 +27,56 @@ const ExchangeBadge = ({ name, css_class }) => `
...
     $('html').addClass('wait');
     if (selector) {
         $(selector).attr('disabled', 'disabled');
     }
 }
 
-function stop_show_loading(selector) {
+function stop_show_loading(selector?: any) {
```
Comment (LefterisJP) [inline ui/user_settings.ts]: isn't a jquery selector a `string` here?

File: ui/user_settings.ts [hunk 26]
```diff
@@ -24,56 +27,56 @@ const ExchangeBadge = ({ name, css_class }) => `
...
     $('html').removeClass('wait');
     if (selector) {
         $(selector).removeAttr('disabled');
     }
 }
 
-function disable_api_entry(selector_text, value) {
+function disable_api_entry(selector_text: any, value: any) {
```
Comment (LefterisJP) [inline ui/user_settings.ts]: isn't a jquery selector a `string` here and also `value` a string as it's just html text to fill in?

File: ui/user_settings.ts [hunk 27]
```diff
@@ -24,56 +27,56 @@ const ExchangeBadge = ({ name, css_class }) => `
...
     $(selector_text).parent().removeClass().addClass('form-group input-group has-success');
-    $(selector_text).attr('disabled', true);
+    $(selector_text).attr('disabled', 'true');
     $(selector_text).val(value);
 }
 
-function enable_api_entry(selector_text) {
+function enable_api_entry(selector_text: any) {
```
Comment (LefterisJP) [inline ui/user_settings.ts]: isn't a jquery selector a `string` here?

File: ui/user_settings.ts [hunk 28]
```diff
@@ -135,10 +138,10 @@ function setup_exchange_callback(event) {
                             }
                             // Exchange removal from backend successful
                             enable_key_entries('', 'exchange');
-                            $('#'+exchange_name+'_badge').remove();
-                            var index = settings.connected_exchanges.indexOf(exchange_name);
-                            if (index == -1) {
-                                throw "Exchange " + exchange_name + "was not in connected_exchanges when trying to remove";
+                            $('#' + exchange_name + '_badge').remove();
+                            let index = settings.connected_exchanges.indexOf(exchange_name.toString());
```
Comment (LefterisJP) [inline ui/user_settings.ts]: why add `toString()` here? Isn't is already a string? It's the value of the `#setup_exchange` textbox.

File: ui/user_settings.ts [hunk 29]
```diff
@@ -171,18 +174,18 @@ function setup_exchange_callback(event) {
                 return;
             }
             // Exchange setup in the backend was successful
-            disable_key_entries('', 'exchange', exchange_name);
-            settings.connected_exchanges.push(exchange_name);
+            disable_key_entries('', 'exchange', exchange_name.toString());
```
Comment (LefterisJP) [inline ui/user_settings.ts]: same here .. why are you adding `toString()` ?

File: ui/user_settings.ts [hunk 30]
```diff
@@ -214,19 +217,19 @@ function fiat_modify_callback(event) {
                 showError('Balance Modification Error', 'Error at modifying ' + currency + ' balance: ' + result['message']);
                 return;
             }
-            if (balance == '') {
-                delete FIAT_BALANCES[currency];
+            if (balance === '') {
+                delete FIAT_BALANCES[currency.toString()];
```
Comment (LefterisJP) [inline ui/user_settings.ts]: same question here. Why add `toString()` ?

File: ui/user_settings.ts [hunk 31]
```diff
@@ -214,19 +217,19 @@ function fiat_modify_callback(event) {
...
         });
 }
 
-function add_user_settings_listeners() {
-    $('#setup_exchange').change(function (event) {
-        if (settings.connected_exchanges.indexOf(this.value) > -1) {
-            disable_key_entries('', 'exchange', this.value);
+export function add_user_settings_listeners() {
+    $('#setup_exchange').change(function (event: any) {
+        if (settings.connected_exchanges.indexOf(this.nodeValue) > -1) {
```
Comment (LefterisJP) [inline ui/user_settings.ts]: Hmm what is the `nodeValue` and what is the different with just `this.value`

File: ui/user_settings.ts [hunk 32]
```diff
@@ -371,16 +374,16 @@ function add_blockchain_account(event) {
...
+    let esc = function ( t: string ) {
         return t
             .replace( /&/g, '&amp;' )
             .replace( /</g, '&lt;' )
             .replace( />/g, '&gt;' )
             .replace( /"/g, '&quot;' );
     };
 
-    return function (d, type, row) {
+    return function (d: any, type: any, row: any) {
```
Comment (LefterisJP) [inline ui/user_settings.ts]: I guess `type` is a `string` here

File: ui/user_settings.ts [hunk 33]
```diff
@@ -396,57 +399,53 @@ let table_data_shortener = function (cutoff_start, keep_length) {
...
     for (let account in eth_accounts) {
         if (eth_accounts.hasOwnProperty(account)) {
             let account_data = eth_accounts[account];
-            let eth_amount = parseFloat(account_data['ETH']);
-            eth_amount = eth_amount.toFixed(settings.floating_precision);
-            let total_usd_value = parseFloat(account_data['usd_value']);
-            total_usd_value = total_usd_value.toFixed(settings.floating_precision);
+            let eth_amount_f = parseFloat(account_data['ETH']);
+            let eth_amount = eth_amount_f.toFixed(settings.floating_precision);
+            let total_usd_value_f = parseFloat(account_data['usd_value']);
```
Comment (LefterisJP) [inline ui/user_settings.ts]: why use different variable names here?

File: ui/user_settings.ts [hunk 34]
```diff
@@ -456,7 +455,7 @@ function recreate_ethchain_per_account_table(eth_accounts) {
     create_ethchain_per_account_table(eth_accounts);
 }
 
-function delete_blockchain_account_row(blockchain, row) {
+function delete_blockchain_account_row(blockchain: any, row: any) {
```
Comment (LefterisJP) [inline ui/user_settings.ts]: `blockchain` here should be a string

File: ui/user_settings.ts [hunk 35]
```diff
@@ -589,7 +588,7 @@ function enable_multiselect() {
 }
 
 
-function add_new_eth_tokens(tokens) {
+function add_new_eth_tokens(tokens: any) {
```
Comment (LefterisJP) [inline ui/user_settings.ts]: If I remember well `tokens` here should be an array of strings (token names)

File: ui/user_settings.ts [hunk 36]
```diff
@@ -616,7 +615,7 @@ function add_new_eth_tokens(tokens) {
     });
 }
 
-function remove_eth_tokens(tokens) {
+function remove_eth_tokens(tokens: any) {
```
Comment (LefterisJP) [inline ui/user_settings.ts]: same here

File: ui/user_settings.ts [hunk 37]
```diff
@@ -677,11 +676,11 @@ function reload_user_settings_tables_if_existing() {
     }
 }
 
-function init_user_settings() {
-    monitor_add_callback('user_settings_query_blockchain_balances', function (result) {
+export function init_user_settings() {
+    monitor_add_callback('user_settings_query_blockchain_balances', function (result: any) {
```
Comment (LefterisJP) [inline ui/user_settings.ts]: result should be an object here

File: ui/userunlock.ts [hunk 38]
```diff
@@ -23,20 +24,20 @@ function verify_userpass(username, password) {
     return true;
 }
 
-function ask_permission(msg, username, password, create_true, api_key, api_secret) {
+function ask_permission(msg: string, username: string, password: string, create_true: any, api_key: string, api_secret: string) {
```
Comment (LefterisJP) [inline ui/userunlock.ts]: `create_true` should be a boolean here

File: ui/userunlock.ts [hunk 39]
```diff
@@ -156,12 +157,12 @@ function unlock_async(username, password, create_true, sync_approval, api_key, a
     return deferred.promise();
 }
 
-function unlock_user(username, password, create_true, sync_approval, api_key, api_secret) {
+function unlock_user(username: string, password: string, create_true: any, sync_approval: any, api_key: string, api_secret: string) {
```
Comment (LefterisJP) [inline ui/userunlock.ts]: `create_true` and `sync_approval` should be a boolean here

File: ui/userunlock.ts [hunk 40]
```diff
@@ -210,24 +211,24 @@ function unlock_user(username, password, create_true, sync_approval, api_key, ap
...
-                        let is_new_user = create_true && api_key == '';
+                        let is_new_user = create_true && api_key === '';
                         load_dashboard_after_unlock(response['exchanges'], is_new_user);
                     });
                     GLOBAL_UNLOCK_DEFERRED = null;
-                }).progress(function(msg){
+                }).progress(function(msg: string) {
                     ask_permission(msg, username, password, create_true, api_key, api_secret);
-                }).fail(function(error){
+                }).fail(function(error: any) {
```
Comment (LefterisJP) [inline ui/userunlock.ts]: shouldn't `error` here be a string?

File: ui/userunlock.ts [hunk 41]
```diff
@@ -210,24 +211,24 @@ function unlock_user(username, password, create_true, sync_approval, api_key, ap
...
-                    self.buttons.ok.action = function () {prompt_sign_in();};
+                    self.buttons.ok.action = function () {prompt_sign_in(); };
                     GLOBAL_UNLOCK_DEFERRED = null;
                 });
         }
     });
 }
 
-function load_dashboard_after_unlock(exchanges, is_new_user) {
+function load_dashboard_after_unlock(exchanges: any, is_new_user: boolean) {
```
Comment (LefterisJP) [inline ui/userunlock.ts]: `exchanges` should be  `Array<string>`

File: ui/utils.ts [hunk 42]
```diff
@@ -0,0 +1,278 @@
...
+import { settings } from './settings';
+import { total_balances_get } from './balances_table';
+import client from './zerorpc_client';
+const {dialog} = require('electron').remote;
+
+
+// Prompt a directory selection dialog and pass selected directory to callback
+// Callback should be a function which accepts a single argument which will be
+// a list of pathnames. The list should only contain 1 entry.
+export function prompt_directory_select_async(callback: any) {
```
Comment (LefterisJP) [inline ui/utils.ts]: can you add the callback type?

File: ui/utils.ts [hunk 43]
```diff
@@ -0,0 +1,278 @@
...
+    dialog.showOpenDialog({
+        title: 'Select a directory',
+        properties: ['openDirectory']
+    }, callback);
+}
+
+
+export function utc_now() {
```
Comment (LefterisJP) [inline ui/utils.ts]: shouldn't this function have a return type?

File: ui/utils.ts [hunk 44]
```diff
@@ -0,0 +1,278 @@
...
+    client.invoke('query_last_balance_save_time', (error, res) => {
+        if (error || res == null) {
+            console.log('Error at periodic client query');
+            return;
+        }
+        settings.last_balance_save = res;
+    });
+}
+
+function _setup_log_watcher(callback: any) {
```
Comment (LefterisJP) [inline ui/utils.ts]: please add the callback type here

File: ui/utils.ts [hunk 45]
```diff
@@ -0,0 +1,278 @@
...
+
+    tail.on('error', function(error: Error) {
+        console.log('TAIL ERROR: ', error);
+    });
+}
+
+// Show an error with TITLE and CONTENT
+// If CALLBACK is given then it should be a callback
+// to call when close is pressed
+export function showError(title: string, content?: string, callback?: any) {
```
Comment (LefterisJP) [inline ui/utils.ts]: callback type here

File: ui/utils.ts [hunk 46]
```diff
@@ -0,0 +1,278 @@
...
+        buttons: {
+            close: function() {
+                return;
+            }
+        }
+    });
+}
+
+// TODO: Remove this/replace with something else. In the case of a huge log hangs the entire app
+export function setup_log_watcher(callback: any) {
```
Comment (LefterisJP) [inline ui/utils.ts]: callback type here

File: ui/utils.ts [hunk 47]
```diff
@@ -0,0 +1,278 @@
...
+    _setup_log_watcher(callback);
+}
+
+export function setup_client_auditor() {
+    if (!client_auditor) {
+        client_auditor = setInterval(periodic_client_query, 60000);
+    }
+}
+
+export function reload_table_currency_val(table: any, colnum: any) {
```
Comment (LefterisJP) [inline ui/utils.ts]: shouldn't `colnum` be a number?

File: ui/utils.ts [hunk 48]
```diff
@@ -0,0 +1,278 @@
...
+    table.rows().invalidate();
+    $(table.column(colnum).header()).text(settings.main_currency.ticker_symbol + ' value');
+    table.draw();
+}
+
+export function reload_table_currency_val_if_existing(table: any, colnum: any) {
```
Comment (LefterisJP) [inline ui/utils.ts]: colnum a number here

File: ui/utils.ts [hunk 49]
```diff
@@ -0,0 +1,278 @@
...
+export function string_capitalize(s: string) {
+    return s && s[0].toUpperCase() + s.slice(1);
+}
+
+export function throw_with_trace(str: string) {
+    let err = new Error();
+    throw new Error(str + '\n' + err.stack);
+}
+
+export function date_text_to_utc_ts(txt: string | number | string[]) {
```
Comment (LefterisJP) [inline ui/utils.ts]: add return type to this function

File: ui/utils.ts [hunk 50]
```diff
@@ -0,0 +1,278 @@
...
+}
+
+interface IctxMenuItems {
+    edit: any;
+    delete: any;
+    sep1: string;
+    quit: any;
+}
+
+export function dt_edit_drawcallback(id: string, edit_fn: any, delete_fn: any) {
```
Comment (LefterisJP) [inline ui/utils.ts]: add types to the `edit_fn` and `delete_fn` callbacks

File: ui/utils.ts [hunk 51]
```diff
@@ -0,0 +1,278 @@
...
+    return function (settings: any) {
```
Comment (LefterisJP) [inline ui/utils.ts]: shouldn't settings be an object here?

File: ui/utils.ts [hunk 52]
```diff
@@ -0,0 +1,278 @@
...
+                }
+
+            },
+            items: ctx_menu_items
+        });
+    };
+}
+
+
+export function unsuggest_element(selector: any) {
```
Comment (LefterisJP) [inline ui/utils.ts]: isn't selector a string?

File: ui/utils.ts [hunk 53]
```diff
@@ -0,0 +1,278 @@
...
+    $(selector).pulsate('destroy');
+    $(selector).removeAttr('style');
+}
+
+export function suggest_element(selector: any, state_to_set: string) {
```
Comment (LefterisJP) [inline ui/utils.ts]: isn't selector a string?

File: ui/utils.ts [hunk 54]
```diff
@@ -0,0 +1,278 @@
...
+        reach: 20,        // how far the pulse goes in px
+        speed: 1000,      // how long one pulse takes in ms
+        pause: 0,         // how long the pause between pulses is in ms
+        glow: true,       // if the glow should be shown too
+        repeat: true,     // will repeat forever if true, if given a number will repeat for that many times
+        onHover: false    // if true only pulsate if user hovers over the element
+    });
+}
+
+export function suggest_element_until_click(selector: any, state_to_set: string) {
```
Comment (LefterisJP) [inline ui/utils.ts]: isn't selector a string?

File: tsconfig.json [hunk 55]
```diff
@@ -0,0 +1,12 @@
+{
```
Comment (LefterisJP) [inline tsconfig.json]: is tsconfig.json supposed to be commited to the repo?

File: tslint.json [hunk 56]
```diff
@@ -0,0 +1,107 @@
+{
```
Comment (LefterisJP) [inline tslint.json]: is tslint.kson supposed to be commited to the repo?

File: tsconfig.json (hunk 55 above)
Comment (StevenJNPearce) [reply tsconfig.json]: Yes, this file contains the configuration for compiling to javascript.

File: ui/utils.ts (hunk 54 above)
Comment (StevenJNPearce) [reply ui/utils.ts]: yes, will fix

File: ui/utils.ts (hunk 53 above)
Comment (StevenJNPearce) [reply ui/utils.ts]: yes, will fix

File: ui/utils.ts (hunk 52 above)
Comment (StevenJNPearce) [reply ui/utils.ts]: yes, will fix

**Comment by StevenJNPearce** [comment]
Ok, Allot of comments. I will have time to go thru this later today.

File: ui/accounting_settings.ts [hunk 57]
```diff
@@ -23,13 +24,13 @@ function populate_ignored_assets() {
...
 
-function ignored_asset_modify_callback(event) {
+function ignored_asset_modify_callback(event: any): void {
     event.preventDefault();
     let button_type = $('#modify_ignored_asset_button').html();
     let asset = $('#ignored_asset_entry').val();
 
     let command = 'add_ignored_asset';
-    if (button_type == 'Remove') {
+    if (button_type === 'Remove') {
```
[the remark this answers is not in the document]
Comment (StevenJNPearce) [reply ui/accounting_settings.ts]: It;s a habit I've picked up from vanilla JS development as good practice, although checking strict doesn't offer any advantage or disadvantage in this case.

File: ui/accounting_settings.ts [hunk 58]
```diff
@@ -52,31 +53,32 @@ function ignored_asset_modify_callback(event) {
...
-            if (command == 'add_ignored_asset') {
+            if (command === 'add_ignored_asset') {
                 $('#ignored_assets_selection').append($('<option>', {
                     value: asset,
                     text: asset,
                     selected: true
                 }));
             } else {
-                $("#ignored_assets_selection option[value='"+asset+"']").remove();
+                $('#ignored_assets_selection option[value=\'' + asset + PageTransitionEvent
```
[the remark this answers is not in the document]
Comment (StevenJNPearce) [reply ui/accounting_settings.ts]: Odd, Don't remember adding this. will remove.

File: ui/definitions.d.ts (hunk 8 above)
Comment (StevenJNPearce) [reply ui/definitions.d.ts]: Provides stub definitions for the JQuery plugins that don't have any available to download.

File: ui/otctrades.ts (hunk 19 above)
Comment (StevenJNPearce) [reply ui/otctrades.ts]: Yes, It just defines the type CURRENT_TRADE should conform to.

File: ui/user_settings.ts (hunk 28 above)
Comment (StevenJNPearce) [reply ui/user_settings.ts]: Right - the type definitions are ignorant as to the type of the underlying HTML element, i.e. `string | number | string[]`. It makes more sense to cast like `$('#setup_exchange').val() as string`, will fix.

File: ui/user_settings.ts (hunk 33 above)
Comment (StevenJNPearce) [reply ui/user_settings.ts]: Because now we have type checking, we can't assign string to number :-)

File: ui/utils.ts (hunk 49 above)
Comment (StevenJNPearce) [reply ui/utils.ts]: compiler can infer type.

File: ui/utils.ts (hunk 51 above)
Comment (StevenJNPearce) [reply ui/utils.ts]: What type of object, what properties does it have? It's an unused variable, should it just be removed?

File: tslint.json (hunk 56 above)
Comment (kelsos) [reply tslint.json]: Yes. This contains rules about the codestyle of the Typescript source files, it will be useful because then you can have a ci task run the lint check and verify that everything is ok.

File: package.json [hunk 59]
```diff
@@ -1,25 +1,32 @@
...
+    "mocha": "^5.1.1",
+    "spectron": "^3.8.0",
+    "tslint": "^5.11.0",
+    "tslint-config-prettier": "^1.14.0",
+    "tslint-config-standard": "^7.1.0",
+    "typescript": "^2.9.2"
+  },
+  "scripts": {
+    "test": "tsc && mocha ui/tests/",
+    "start": "tsc && electron ."
```
Comment (kelsos) [inline package.json]: You could also add a script to run tslint `"lint": "tslint -c tslint.json --project tsconfig.json 'ui/**/*.ts'"` here.

**Comment by StevenJNPearce** [comment]
@LefterisJP I will be completely honest here, I lost interest in finishing this despite there being little remaining to complete it. I've gone ahead and done 'stop work' on gitcoin.

**Comment by LefterisJP** [comment]
@StevenJNPearce Hey man how come? I have just been waiting for you to implement the comments from the PR review and rebased on master to go ahead with merging.

I don't think it's that much work.

**Comment by StevenJNPearce** [comment]
Motivation is elusive... wish I had that figured out better. 

Can you define the interface query results should conform to, e.g. `monitor_add_callback('user_settings_query_blockchain_balances', function (result: any)` What does result look like? What properties does it have?
